category/category.go

```diff
@@ -0,0 +1,269 @@
+	"github.com/almighty/almighty-core/gormsupport"
+	"github.com/almighty/almighty-core/log"
+	"github.com/jinzhu/gorm"
+	errs "github.com/pkg/errors"
+	uuid "github.com/satori/go.uuid"
+)
+
+// Defines "type" string to be used while validating jsonapi spec based payload
+const (
+	APIStringType = "categories"
+)
+
+// String constants for system categories
+const (
+	PlannerRequirements = "planner.requirements"
+	PlannerPortfolio    = "planner.portfolio"
+)
+
+// WARNING: Do not change these UUIDs!
+// System defined categories
+var (
+	PlannerRequirementsID = uuid.FromStringOrNil("b51ccf69-d574-41c5-b738-4a69265129d1") // "planner.requirements"
+	PlannerPortfolioID    = uuid.FromStringOrNil("0625e4bf-122a-4c1c-8ccd-e0995ef31974") // "planner.portfolio"
+)
+
+// Category describes a single category
+type Category struct {
+	gormsupport.Lifecycle
+	ID   uuid.UUID `sql:"type:uuid default uuid_generate_v4()" gorm:"primary_key"` // This is the ID PK field
+	Name string
+}
+
+// TableName overrides the table name settings in Gorm to force a specific table name
+// in the database.
+func (category *Category) TableName() string {
+	return "categories"
+}
+
+// WorkItemTypeCategoryRelationship describes relationship between a category and a workitemtype.
+type WorkItemTypeCategoryRelationship struct {
+	gormsupport.Lifecycle
+	ID             uuid.UUID `sql:"type:uuid default uuid_generate_v4()" gorm:"primary_key"` // This is the ID PK field
+	CategoryID     uuid.UUID `sql:"type:uuid"`
+	WorkItemTypeID uuid.UUID `sql:"type:uuid"`
+}
+
+// TableName overrides the table name settings in Gorm to force a specific table name
+// in the database.
+func (m *WorkItemTypeCategoryRelationship) TableName() string {
+	return "work_item_type_categories"
+}
+
+// Ensure fields implements the Equaler interface
+var _ convert.Equaler = Category{}
+var _ convert.Equaler = (*Category)(nil)
+
+// Equal returns true if two Category objects are equal; otherwise false is returned.
+func (category Category) Equal(u convert.Equaler) bool {
+	other, ok := u.(Category)
+	if !ok {
+		return false
+	}
+	if !uuid.Equal(category.ID, other.ID) {
+		return false
+	}
+	if !category.Lifecycle.Equal(other.Lifecycle) {
+		return false
+	}
+	if category.Name != other.Name {
+		return false
+	}
+	return true
+}
+
+// Repository encapsulates storage and retrieval of categories
+type Repository interface {
+	Create(ctx context.Context, category *Category) (*Category, error)
+	LoadCategoryFromDB(ctx context.Context, id uuid.UUID) (*Category, error)
+	List(ctx context.Context) ([]*Category, error)
+	CreateRelationship(ctx context.Context, relationship *WorkItemTypeCategoryRelationship) error
+	LoadWorkItemTypeCategoryRelationship(ctx context.Context, workitemtypeID uuid.UUID, categoryID uuid.UUID) (*WorkItemTypeCategoryRelationship, error)
+	LoadAllRelationshipsOfCategory(ctx context.Context, categoryID uuid.UUID) ([]*WorkItemTypeCategoryRelationship, error)
```

Comment:
Yes, that is fine but I still wonder if the only link to a category can be between a work item type and a category. Can't link types also belong to a category? I'm not sure to be honest.